### ISSUE TITLE
False Positive: Trivial delegation and `Comparable` exceptions are still treated as `equals`/`hashCode` pairing violations.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6536]
# False Positive: Trivial delegation and `Comparable` exceptions are still treated as `equals`/`hashCode` pairing violations.

Version
7.22.0

## Checker
- Checker id: `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`
- Checker description: Detects when a class (including anonymous classes and records) overrides either `equals` or `hashCode` but not both, unless the class implements `Comparable`.

## Description of the false positive
`diff/output` marks 3 false positives for `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are abstract declarations, delegation, suppression or annotations.

### `NegCase3.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/neg/NegCase3.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. Reporting this as if the class had already committed to a concrete contract is too aggressive. That matches the benchmark's stated intent: A class with abstract equals method should not be flagged as missing hashCode.

```java
// A class with abstract equals method should not be flagged as missing hashCode.
package scensct.core.neg;

public abstract class NegCase3 {
    public abstract boolean equals(Object obj);
}
```

### `NegCase4.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/neg/NegCase4.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. Reporting this as if the class had already committed to a concrete contract is too aggressive. That matches the benchmark's stated intent: A class with abstract hashCode method should not be flagged as missing equals.

```java
// A class with abstract hashCode method should not be flagged as missing equals.
package scensct.core.neg;

public abstract class NegCase4 {
    public abstract int hashCode();
}
```

### `NegCase3.java`
Source: `output/errorprone/EqualsHashCode/src/main/java/scensct/core/neg/NegCase3.java`

The only interesting behavior here is an `equals(Object)` override that immediately delegates to `super.equals(obj)`. That does not establish a new equality relation; it preserves `Object`'s identity-based behavior. That matches the benchmark's stated intent: A Java class that declares an equals method with trivial delegation to Object.equals should not be flagged as overriding equals without hashCode.

```java
// A Java class that declares an equals method with trivial delegation to Object.equals should not be flagged as overriding equals without hashCode.
package scensct.core.neg;

public class NegCase3 {
    private String data;
    
    public NegCase3(String d) {
        this.data = d;
    }
    
    // Trivial equals that just delegates to super.equals
    @Override
    public boolean equals(Object obj) {
        return super.equals(obj);  // Single statement delegating to Object.equals
    }
}
```

## Cause analysis
The rule is documented as follows: Detects when a class (including anonymous classes and records) overrides either `equals` or `hashCode` but not both, unless the class implements `Comparable`.

In this set the problem shows up around abstract declarations, delegation, suppression or annotations. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.

**Comment by oowekyala** [comment]
This doesn't look like a problem to me. If you declare equals abstract it's to force the implementors to implement it. Unless you also declare hashCode abstract implementors might forget about overriding that one so it is a problem.

The last example is also a case which is missing the point of the rule. It cannot check whether you are defining a new equivalence relationship when implementing equals, or whether it is consistent with hashcode. It can only remind you that you need to care about the other when you implement the one, whatever the implementation is.
